xz-bootstrap: cleanup install path

catap opened the pull request:
#### Description

As part of https://github.com/macports/macports-ports/pull/13995 I've removed `libexec/libcxx-bootstrap`.

Here simple move to where `xz-bootstrap` is installed.

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
macOS x.y
Xcode x.y

###### Verification <!-- (delete not applicable items) -->
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

jmroot commented:
Why would this be desirable? Doesn't just removing dependencies mean that things will be used opportunistically?

catap commented:
@jmroot all this bootstrap ports are required to build cmake and clang on very old system which allows to build libcxx.

All this dependencies are optional, and it can be build and use without when on clean system.

The goal for `-bootstrap` build is built something with as little dependencies as possible.

jmroot commented:
Users don't generally have clean systems. In any case that only explains removing some of the dependencies, not all the other changes.

catap commented:
@jmroot here a few changes.

1. Moving `xz-bootstrap` and `python27-bootstrap` from `libexec/libcxx-bootsrap`.

Why? To keep things clean. Right now `cmake-bootsrap` is installed to `libexec/cmake-bootsrap`; brand new `gcc10-bootsrap` is installed to `libexec/gcc10-bootstrap` and the same happens with `clang-11-boostrap` which hasn't been yet merged.

2. Remove dependencies from `python27-bootstrap`.

`python27-bootstrap` has a few dependencies and one of them is openssl. Python27 is old enough and can be built with old openssl from macOS 10.6 and early.

I feel that this dependency was added as a sort of side effect when openssl PG was introduced.

Thus, this port is *build* requirements for clang and cmake, this port shouldn't be used in normal life. Only to bootstrap them.

So, here, I simple removed all this unnecessary dependency and if it picks something from installed system, well, it won't make any harm. is it?

3. Mark some ports obsolete.

All this ports was used via `python27-bootstrap` and as soon as I've cleaned up this dependency... well... we don't need them anymore.

Do you see any use case of `gperf-bootstrap` for example?

kencu commented:
I'd just like add a short note of support for what @catap is doing here, and all the work and attention he is putting into this issue of late. 

Trimming down the dep tree for the bootstrap ports if possible, especially python27, does make the bootstrap setup considerably simpler. Certainly we have to then make sure that somehow the deps aren't just found opportunistically instead, as that would just cause havoc.

Obsoleting some of these bootstrap ports at this point in time is very neat-and-tidy but might be a bit premature -- we could leave them sit for a bit to be sure they aren't needed.

jmroot commented:
> 1. Moving `xz-bootstrap` and `python27-bootstrap` from `libexec/libcxx-bootsrap`.
> 
> Why? To keep things clean. Right now `cmake-bootsrap` is installed to `libexec/cmake-bootsrap`; brand new `gcc10-bootsrap` is installed to `libexec/gcc10-bootstrap` and the same happens with `clang-11-boostrap` which hasn't been yet merged.

I don't think that's a good enough reason to make everyone reinstall these ports.

> 2. Remove dependencies from `python27-bootstrap`.
> 
> `python27-bootstrap` has a few dependencies and one of them is openssl. Python27 is old enough and can be built with old openssl from macOS 10.6 and early.
> 
> I feel that this dependency was added as a sort of side effect when openssl PG was introduced.
> 
> Thus, this port is _build_ requirements for clang and cmake, this port shouldn't be used in normal life. Only to bootstrap them.
> 
> So, here, I simple removed all this unnecessary dependency and if it picks something from installed system, well, it won't make any harm. is it?

In my experience, undeclared dependencies being picked up causes many problems. So I will only accept this if you can show that the system libs will always be used.

> 3. Mark some ports obsolete.
> 
> All this ports was used via `python27-bootstrap` and as soon as I've cleaned up this dependency... well... we don't need them anymore.

Of course, obsoleting them if they have no dependents makes sense.

jmroot commented on devel/gperf/Portfile:
```diff
@@ -27,18 +26,14 @@ configure.args  --infodir=${prefix}/share/info
+    PortGroup   obsolete 1.0
+    revision    [ expr ${revision} + 1 ]
+
+    #------------------------------------------------------------------------------
```
Ascii art to highlight this comment isn't necessary.

jmroot commented on devel/ncurses/Portfile:
```diff
@@ -1,5 +1,6 @@
-PortSystem 1.0
```
clang_dependency portgroup again

jmroot commented on devel/pkgconfig/Portfile:
```diff
@@ -1,7 +1,6 @@
 # -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
 
 PortSystem          1.0
-PortGroup           clang_dependency 1.0
```
and again

jmroot commented on lang/llvm-3.4/Portfile:
```diff
@@ -77,15 +75,12 @@ if {${subport} eq "llvm-${llvm_version}"} {
-
-    configure.cppflags-prepend  -I${bootstrap_prefix}/include
-    configure.ldflags-prepend   -L${bootstrap_prefix}/lib
-    configure.env-append        PATH=${bootstrap_prefix}/bin:$::env(PATH)
```
I assume you checked that this works right without the PATH change?

catap replied:
@jmroot yes, I've build it it.

jmroot commented on lang/python27/Portfile:
```diff
@@ -68,33 +68,30 @@ openssl.branch      1.1
-    # no need to be 'port select'able
+
+    depends_build-delete    port:pkgconfig
+    # OpenSSL PG adds dependency via callback, clena it the same way :)
```
Better to avoid including the portgroup in the first place if it shouldn't be used.

catap replied:
Good idea, indeed.

jmroot commented on textproc/libiconv/Portfile:
```diff
@@ -2,7 +2,6 @@
 
 PortSystem              1.0
 PortGroup               muniversal 1.0
-PortGroup               clang_dependency 1.0
```
clang_dependency again

kencu commented on devel/gperf/Portfile:
```diff
@@ -1,7 +1,6 @@
 # -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
 
-PortSystem 1.0
-PortGroup  clang_dependency 1.0
```
the idea is that it is about to disappear, and be replaced by something new.

system roots -> gcc10-bootstrap -> all deps needed to build clang11-bootstrap -> everything

the order of operations here has to match, however

kencu replied:
perhaps the something new will still need to be in a revised clang dependency portgroup, however. I haven’t looked deeply into this part at all yet, as right now the process is still stuck at getting gcc10-bootstrap built right.

kencu replied:
right now they do -- but the idea will be to move all of them (all possible anyway) to build with clang11-bootstrap.

kencu replied:
and someone could possibly (possibly) dump all the older ones, as none of them would be needed any longer.

kencu replied:
That is all to be sorted out once this other stuff (gcc10-bootstrap and clang11-bootstrap work). 

The basic idea is that all the clang deps will build with clang11-bootstrap, and the clang_dependency tangle that makes them all build with ancient clang versions will disappear (as will, hopefully, all the ancient clang versions).

kencu replied:
@catap -- I'm not sure where this PR is headed. You might just close it? We have to get the other stuff sorted out and come up with a coherent plan first.

jmroot replied:
> The basic idea is that all the clang deps will build with clang11-bootstrap

That is exactly what the portgroup ensures, just with a different clang port. So I maintain that removing it is not appropriate in this PR. (And most likely it should never be removed, just modified.)

catap replied:
yeah, it was defently too early. Too early.

jmroot replied:
> I'm not sure where this PR is headed. You might just close it? We have to get the other stuff sorted out and come up with a coherent plan first.

If you need to rethink things that's fine. But please don't interpret criticism of the specifics of a proposal as outright rejection of the entire idea.

catap commented:
@kencu, @jmroot I'm back. Sorry, I have quite a mess in my real life which required a lot of attention.

I just made the force push to this PR which should address all question.

Let me summarise it here to make things simpler.

1. Removing clang_dependency PG was a mistake. I've tried to delivere everything in one huge PR but @kencu raised a point that it should be split and I've forgot to clean it up here. Should be better now.

2. Dependencies. I've implemented the same logic that I've used inside gcc10-bootstrap port to sterilise everything to not depends on anything from macports. I've also tested it with installing all dependencies from non bootstrap version, build from scratch bootstrap port, remove everything and run rev-dep.

3. Prefix was set correctly and ascii art was also removed.

jmroot commented on devel/pkgconfig/Portfile:
```diff
@@ -38,21 +38,11 @@ set docdir          ${prefix}/share/doc/${name}
 
 # This port is used by clang-3.4 to bootstrap libcxx
 subport ${name}-bootstrap {
-    set myprefix ${prefix}/libexec/libcxx-bootstrap
```
Similar to xz, all use of `myprefix` should be removed.

catap replied:
done

jmroot commented on lang/llvm-3.4/Portfile:
```diff
@@ -77,15 +75,12 @@ if {${subport} eq "llvm-${llvm_version}"} {
 
 if {${os.platform} eq "darwin" && ${os.major} < 11 && ${cxx_stdlib} eq "libc++"} {
-    configure.cxx_stdlib    libstdc++
+    revision                    [ expr ${revision} + 1 ]
```
Expressions should be brace-quoted. Having a different revision depending on the platform is a bit problematic anyway though.

catap replied:
Let bump whole revision when.

jmroot commented on lang/python27/Portfile:
```diff
@@ -64,40 +63,65 @@ depends_lib         port:bzip2 \
-    prefix ${prefix}/libexec/libcxx-bootstrap
+    revision                [ expr ${revision} + 1 ]
+
+    use_xz                  no
```
Unfortunately I don't think this works the way you would expect.

catap replied:
I expect that it doesn't try to install `xz` by dependencies for extract phase.

Instead it uses `xz-bootstrap` and provided command to extract an archive.

Also, as it works. Have I missed something?

jmroot replied:
In any port that currently sets `use_xz yes` (autoconf for example), try adding `use_xz no` after that. You'll see that xz is still listed in the extract dependencies.

So this line does nothing, and it's the overwriting of `depends_extract` that removes the xz dep.

jmroot commented on lang/python27/Portfile:
```diff
@@ -64,40 +63,65 @@ depends_lib         port:bzip2 \
+    compiler.cpath
+    compiler.library_path
+
+    configure.cflags
+    configure.cxxflags
+    configure.cppflags
+    configure.optflags
+    configure.ldflags
+
+    configure.universal_cflags
+    configure.universal_cxxflags
+    configure.universal_cppflags
+    configure.universal_ldflags
+    configure.universal_args
```
Same comment as for xz.

jmroot commented on lang/python27/Portfile:
```diff
@@ -64,40 +63,65 @@ depends_lib         port:bzip2 \
-                        ac_cv_lib_intl_textdomain=no \
-                        ac_cv_search_bind_textdomain_codeset=no
+
+    configure.args          ac_cv_header_libintl_h=no \
+                            ac_cv_lib_intl_textdomain=no \
+                            ac_cv_search_bind_textdomain_codeset=no
```
Is this still necessary and appropriate? Shouldn't using the system gettext be OK?

catap replied:
@jmroot it was here before, and I can try to remove it, but it may broke the port somewhere. I prefer to keep it here. Do you suggest to remove it?

jmroot replied:
If you're worried about changes breaking the port, you probably shouldn't make all these other changes. :)

This was only here to prevent gettext from being found in `$prefix` (b68b114), so it should no longer be needed with the flags cleared of references to `$prefix`.

jmroot commented on textproc/libiconv/Portfile:
```diff
@@ -39,18 +39,11 @@ set myprefix ${prefix}
 
 # This port is used by clang-3.4 to bootstrap libcxx
 subport ${name}-bootstrap {
-    set myprefix ${prefix}/libexec/libcxx-bootstrap
```
Use of `myprefix` needs to be cleaned up in this Portfile too.

catap replied:
Done

catap commented on devel/gperf/Portfile:
```diff
@@ -27,18 +27,11 @@ configure.args  --infodir=${prefix}/share/info
+    revision    1
+
+    # Obsolete Date: 2022-01-17
+    # This was used only by libiconv-bootstrap, no replacment here.
```
got it

catap commented on devel/gperf/Portfile:
```diff
@@ -1,7 +1,7 @@
-PortSystem 1.0
-PortGroup  clang_dependency 1.0
+PortSystem      1.0
+PortGroup       clang_dependency 1.0
```
reverted

catap commented on archivers/xz/Portfile:
```diff
@@ -47,10 +47,41 @@ if {${subport} eq ${name}} {
+    configure.env-append    PATH=/usr/bin:/bin:/usr/sbin:/sbin
+    build.env-append        PATH=/usr/bin:/bin:/usr/sbin:/sbin
+
+    # sterilize dependencies
+    depends_build
+    depends_lib
+    depends_run
```
@jmroot that adds a few if. Right now it keeps everything related to bootstrap in bootstrap subport.

jmroot replied:
Fair enough.

catap commented on archivers/xz/Portfile:
```diff
@@ -47,10 +47,41 @@ if {${subport} eq ${name}} {
 # This port is used by clang-3.4 to bootstrap libcxx
 subport ${name}-bootstrap {
-    revision                1
-    set my_prefix ${prefix}/libexec/libcxx-bootstrap
```
done

jmroot replied:
There are still a few `my_prefix` occurrences.

jmroot commented:
The adding of flags for db48 in the python27 Portfile needs to be prevented for the bootstrap port. (This was a pre-existing oversight.)

catap commented on archivers/xz/Portfile:
```diff
@@ -47,10 +47,41 @@ if {${subport} eq ${name}} {
+    configure.universal_cxxflags
+    configure.universal_cppflags
+    configure.universal_ldflags
+    configure.universal_args
```
cleaned up

kencu replied:
that's on me. I am never 100% sure what is going to slip in where, and went a bit overboard to purify the build I guess.

catap replied:
@kencu to be honest I agree with you paranoia and personally I'm also a bit paranoid and tried to except side effects here :)

catap commented:
@jmroot if I not mistaken I've addressed all your comments. May I ask you to re-review it?

jmroot commented on archivers/xz/Portfile:
```diff
@@ -47,10 +47,29 @@ if {${subport} eq ${name}} {
+    compiler.cpath
+    compiler.library_path
+
+    configure.cflags
+    configure.cxxflags
```
`cflags` and `cxxflags` shouldn't be cleared.

jmroot commented on devel/pkgconfig/Portfile:
```diff
@@ -67,7 +52,7 @@ configure.args      --disable-silent-rules \
 
 if {$subport eq "${name}-bootstrap"} {
     # search bootstrap prefix and then normal prefix
-    configure.args-append   --with-pc-path=${myprefix}/lib/pkgconfig:${myprefix}/share/pkgconfig:${prefix}/lib/pkgconfig:${prefix}/share/pkgconfig
+    configure.args-append   --with-pc-path=${prefix}/lib/pkgconfig:${prefix}/share/pkgconfig:${prefix}/lib/pkgconfig:${prefix}/share/pkgconfig
```
This branch of the if statement will never be taken, since pkgconfig-bootstrap no longer exists.

jmroot commented on lang/python27/Portfile:
```diff
@@ -61,43 +63,48 @@ depends_lib         port:bzip2 \
 subport ${name}-bootstrap {
-    set stdprefix ${prefix}
-    prefix ${prefix}/libexec/libcxx-bootstrap
+    revision                [ expr ${revision} + 1 ]
```
Remember to [brace your expressions](https://wiki.tcl-lang.org/page/Brace+your+expr%2Dessions). :)

jmroot reviewed (changes requested):
Getting close.

catap commented:
@jmroot here it is

kencu reviewed (commented):
Thanks for asking for review, will leave this with @jmroot
